### DISCUSSION TITLE
How to make sure docker daemon sucessfully started even through PID file existed

### DISCUSSION THREAD

**Discussion by botieking98** [moby/moby#46992] [Q&A]
### Description

Now docker daemon will start failed with error `process with PID xxx is still running` when `/var/run/docker.pid` file existed and process is alive.

The associated code is below:
```go
// Read reads the "PID file" at path, and returns the PID if it contains a
// valid PID of a running process, or 0 otherwise. It returns an error when
// failing to read the file, or if the file doesn't exist, but malformed content
// is ignored. Consumers should therefore check if the returned PID is a non-zero
// value before use.
func Read(path string) (pid int, err error) {
	pidByte, err := os.ReadFile(path)
	if err != nil {
		return 0, err
	}
	pid, err = strconv.Atoi(string(bytes.TrimSpace(pidByte)))
	if err != nil {
		return 0, nil
	}
	if pid != 0 && process.Alive(pid) {
		return pid, nil
	}
	return 0, nil
}

// Write writes a "PID file" at the specified path. It returns an error if the
// file exists and contains a valid PID of a running process, or when failing
// to write the file.
func Write(path string, pid int) error {
	if pid < 1 {
		// We might be running as PID 1 when running docker-in-docker,
		// but 0 or negative PIDs are not acceptable.
		return fmt.Errorf("invalid PID (%d): only positive PIDs are allowed", pid)
	}
	oldPID, err := Read(path)
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	if oldPID != 0 {
		return fmt.Errorf("process with PID %d is still running", oldPID)
	}
	return os.WriteFile(path, []byte(strconv.Itoa(pid)), 0o644)
}
``` 

### Reproduce

We should do this:
1. Let /var/run/docker.pid file redundant
2. Reboot machine
3. Start docker and PID of docker daemon is same to PID file content

docker daemon failed started occasionally.

### Expected behavior

docker daemon started sucessfully even through PID file existed and is same to PID of dockerd

### docker version

```bash
ALL version
```

### docker info

```bash
ALL version
```

**Comment by thaJeztah** [comment]
This behavior is intentional, and is to prevent multiple instances from the daemon to be started (running multiple instances of the daemon that interact with the same data-store can cause corruption of data).

On a normal shutdown of the docker daemon service, the PID-file should be removed;

```bash
cat /var/run/docker.pid
101388

systemctl stop docker.service
Warning: Stopping docker.service, but it can still be activated by:
  docker.socket

cat /var/run/docker.pid
cat: /var/run/docker.pid: No such file or directory
```


> Reboot machine

Did this lead to an unclean shutdown? Is the docker daemon running as a systemd unit in your setup, or is it started through other means? (See above; the PID-file should be removed on a clean shut-down).

**Comment by thaJeztah** [comment]
As I don't think there's a bug here, I'll move this to a discussion

**Comment by dbuzatto** [comment]
Hi! This is a known race condition related to PID reuse after reboot.

**Root cause**

After a reboot, the Linux kernel recycles PIDs. If any process 
(e.g. systemd, a kernel thread, or any early-boot process) happens 
to get the same PID that was stored in `/var/run/docker.pid`, the 
`process.Alive()` check returns `true` — even though that process 
has nothing to do with Docker.

The current code only checks if *a* process with that PID is running, 
not if it's *actually dockerd*:

if pid != 0 && process.Alive(pid) {
    return pid, nil  // assumes it's still dockerd — but it might not be!
}

**Why it's occasional**

PID assignment after boot isn't deterministic. It only fails when 
the recycled PID matches the one in the file, which is why this 
doesn't happen on every reboot.

**Workaround (operational fix)**

The cleanest workaround is to use systemd to remove the PID file 
before starting Docker. Add this to your docker.service override:

[Service]
ExecStartPre=/bin/rm -f /var/run/docker.pid

Run:
systemctl edit docker

Paste the block above, save, then:
systemctl daemon-reload && systemctl restart docker

This ensures the stale PID file is always cleaned up before 
dockerd tries to start.

**Alternative: tmpfs for /var/run**

On most modern Linux distros, `/var/run` is already a tmpfs mount, 
meaning it's wiped on every reboot. If you're hitting this issue, 
double-check:

mount | grep /run

If it's not tmpfs, that's likely your real problem — the PID file 
is persisting across reboots when it shouldn't.

**Proper fix (code level)**

The `Write()` function should additionally verify that the process 
found at the stored PID is actually dockerd, for example by checking 
`/proc/<pid>/comm` or `/proc/<pid>/exe`:

// pseudocode
if oldPID != 0 && isActuallyDockerDaemon(oldPID) {
    return fmt.Errorf("process with PID %d is still running", oldPID)
}
// otherwise: stale PID from a recycled process, safe to overwrite

This would make the check resilient to PID reuse after reboot.

Hope this helps! The ExecStartPre workaround should solve it 
immediately while a proper fix isn't merged.
